### ISSUE TITLE
Support `hostname` in LB Service status

### ISSUE THREAD

**Issue by youngnick** [cilium/cilium#23101]
Currently, Cilium will add rules for any addresses in the `status.loadBalancer.ingress.ip` stanza of relevant Loadbalancer Services. However, it's valid for Loadbalancer Services to use `status.loadBalancer.ingress.hostname` instead (AWS LB Services do this). In this case, we should be turning the hostname into IP addresses and adding those in the same ways that we add LB Service IPs currently.

This is extra complicated because we'll need to maintain a watch on the hostname and update the IPs when the TTLs expire, which will probably require interaction with the DNS proxy.

This issue covers figuring out what the best way forward is here, and then implementing something.

**Comment by youngnick** [comment]
I think this is still important for us to get to at some point.

**Comment by dal13002** [comment]
@youngnick I am currently looking for this exact feature as this avoids hairpinning traffic (ie cluster -> AWS NLB -> cluster). You would have to maintain an a watch on the hostname and update the IPs when the TTLs expire for certain type of load balancers (AWS ALB), but not others (AWS NLB), which have static IPs. I guess just to be safe you could watch everything if the status field has the hostname. If it is easier to add an annotation allowing the user to confirm hostname has static ip and they want this, we could resolve it -> handle it same way as `status.loadBalancer.ingress.ip`. I think there is a lot of value in that specifically with cost savings / lower latency. What are your thoughts on that?

**Comment by youngnick** [comment]
I don't think that having Cilium manage this DNS lookup for you is a good idea, as that would require polling DNS every time the TTL expires, which for this sort of record is often _very_ frequently, on the order of once every 5 seconds.

In a cluster with many Gateways of this type, there is a big risk of overloading the API server by constantly updating the addresses in the Gateway's status.

When I said "support `hostname`" in this issue earlier, I mean literally just copying the `hostname` value from the Loabalancer Service into the Gateway. Currently, Cilium does not support that.

**Comment by dal13002** [comment]
Okay, I misunderstood the issue. Could you explain what you mean by "This is extra complicated because we'll need to maintain a watch on the hostname and update the IPs when the TTLs expire, which will probably require interaction with the DNS proxy." from the original issue?

**Comment by youngnick** [comment]
Okay, I made some other updates here and in the issue that were just wrong.

We actually shouldn't be dereferencing these addresses at all when handling Services, precisely because of this reason.

The only cases we have are:

* Cilium is using LB-IPAM to hand out addresses, in this case the addresses should be in the status.
* Cilium is using an upstream cloud provider like AWS for LB Services, that puts an upstream `hostname` into the status. We leave this as it is.

For _consumers_ of this status info, like Gateway API, we should just use whatever is in the LB Service, without doing our own lookups.

To be honest, I think I may have been mistaken in logging this issue originally. So I'm going to close it.

Sorry for the confusion @dal13002.
